components/camera/replaypcd/replaypcd_utils_test.go

```diff
@@ -0,0 +1,207 @@
+
+	newFileNum, err := getNextDataAfterFilter(filter, last)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get point cloud data in gzip compressed format
+	path := filepath.Clean(artifact.MustPath(fmt.Sprintf(datasetDirectory, newFileNum)))
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var dataBuf bytes.Buffer
+	gz := gzip.NewWriter(&dataBuf)
+	gz.Write(data)
+	gz.Close()
+
+	// Construct response
+	binaryData := &datapb.BinaryData{
+		Binary:   dataBuf.Bytes(),
+		Metadata: &datapb.BinaryMetadata{},
+	}
+
+	resp := &datapb.BinaryDataByFilterResponse{
+		Data: []*datapb.BinaryData{binaryData},
+		Last: fmt.Sprint(newFileNum),
+	}
+	return resp, nil
+}
+
+// createMockCloudDependencies creates a mockDataServiceServer and rpc client connection to it which is then
+// stored in a mockCloudConnectionService.
+func createMockCloudDependencies(ctx context.Context, t *testing.T, logger golog.Logger) (resource.Dependencies, func() error) {
+	listener, err := net.Listen("tcp", "localhost:0")
+	test.That(t, err, test.ShouldBeNil)
+	rpcServer, err := rpc.NewServer(logger, rpc.WithUnauthenticated())
+	test.That(t, err, test.ShouldBeNil)
+
+	test.That(t, rpcServer.RegisterServiceServer(
+		ctx,
+		&datapb.DataService_ServiceDesc,
+		&mockDataServiceServer{},
+		datapb.RegisterDataServiceHandlerFromEndpoint,
+	), test.ShouldBeNil)
+
+	go rpcServer.Serve(listener)
+
+	conn, err := viamgrpc.Dial(ctx, listener.Addr().String(), logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	r := &inject.Robot{}
+	rs := map[resource.Name]resource.Resource{}
+	rs[cloud.InternalServiceName] = &mockCloudConnectionService{
+		Named: cloud.InternalServiceName.AsNamed(),
+		conn:  conn,
+	}
+
+	r.MockResourcesFromMap(rs)
+
+	return resourcesFromDeps(t, r, []string{cloud.InternalServiceName.String()}), rpcServer.Stop
+}
+
+// createNewReplayPCDCamera will create a new replay_pcd camera based on the provided config with either
+// a valid or invalid data client.
+func createNewReplayPCDCamera(ctx context.Context, t *testing.T, replayCamCfg *Config, validDeps bool,
+) (camera.Camera, func() error, error) {
+	logger := golog.NewTestLogger(t)
+
+	var resources resource.Dependencies
+	var closeRPCFunc func() error
+	if validDeps {
+		resources, closeRPCFunc = createMockCloudDependencies(ctx, t, logger)
+	}
+
+	cfg := resource.Config{ConvertedAttributes: replayCamCfg}
+	cam, err := newPCDCamera(ctx, resources, cfg, logger)
+
+	return cam, closeRPCFunc, err
+}
+
+var _ = cloud.ConnectionService(&mockCloudConnectionService{})
+
+// mockCloudConnectionService creates a mocked version of a cloud connection service.
+type mockCloudConnectionService struct {
+	resource.Named
+	resource.AlwaysRebuild
+	conn rpc.ClientConn
+}
+
+// AcquireConnection returns a connection to the rpc server stored in the mockCloudConnectionService object.
+func (noop *mockCloudConnectionService) AcquireConnection(ctx context.Context) (string, rpc.ClientConn, error) {
```

Comment:
I'd like to see a test written where this fails as well as BinaryDataByFilter so that we have good coverage around those fail points

Reply:
I have added a test, on line 52 that will force an error response from the call to AcquireConnection and check to make sure it was received. 

Tests for failure modes on BinaryDataByFilter have already been tested by the tests checking if errEndOfDataset was produced when no files were available to be read.